File: Data Structures & Algorithms/generate-parentheses/submission-0.cpp
class Solution {
public:
    bool isValid(string s){
        stack<char> st;
        for(int i = 0 ; i< s.length(); i++){
            if(s[i] == '('){
                st.push(s[i]);
            }
            else{
                if(st.empty())
                return false;
                else
                st.pop();
            }
        }
        return st.size() == 0;
    }
    void solve(string &s, vector<string> &ans, int i,int n){
        if(i == n){
            if(isValid(s))
            ans.push_back(s);
            return;
        }
        unordered_set<char> st;
        for(int j = i; j < n; j++){
            if(st.count(s[j])) continue;

            st.insert(s[j]);
            swap(s[i], s[j]);
            solve(s, ans, i+1, n);
            swap(s[i], s[j]);
        }
        return;
    }
    vector<string> generateParenthesis(int n) {
        string s = "";
        for(int i = 0 ; i< n;  i++){
            s += "()";
        }
        vector<string> ans;
        solve(s, ans, 0, n*2);
        return ans;
    }
};

File: Data Structures & Algorithms/longest-common-subsequence/submission-0.java
class Solution {
    private int solve(String s1, String s2, int m, int n, int [][] dp){
        if(m == 0 || n == 0) return 0;
        if(dp[m-1][n-1] != 0){
            return dp[m-1][n-1];
        }
        int ans = 0;
        if(s1.charAt(m-1) == (s2.charAt(n-1))){
            ans = 1 + solve(s1, s2, m-1, n-1, dp);
        }
        else {
            ans = Math.max(solve(s1, s2, m-1, n, dp), solve(s1, s2, m, n-1, dp));
        }
        return dp[m-1][n-1] = ans;
    }
    public int longestCommonSubsequence(String s1, String s2) {
        int m = s1.length(), n = s2.length();
        int [][] dp = new int [m][n];
        for(int i= 0; i< m; i++){
            for(int j = 0 ;j< n; j++){
                if(i == 0 || j == 0){
                    dp[i][j] = 0;
                }
                else{break;}
            }
        }
        return solve(s1, s2, m, n, dp);
    }
}
